Sources/TensorFlow/Core/LazyTensorTrace.swift

```diff
@@ -0,0 +1,160 @@
+    // inputs will be "placeholder" nodes.
+    var inputs: [LazyTensorOperation] = []
+    var inputValues: [TFETensorHandle] = []
+    var operations: [LazyTensorOperation] = []
+    var outputs: [LazyTensorOperation] = []
+    var originalOutputs: [LazyTensorOperation] = []
+
+    init(_ lazyOp: LazyTensorOperation) {
+        // TODO: We only pick operations on which `lazyOp` depends on. Note that
+        // there may be other live tensors that could also be materialized at
+        // this time. e.g.,
+        //   x = a + b
+        //   y = x + c
+        // For `x`, only `a + b` is extracted. One optimization is to also include
+        // `y = x + c` into the trace so that we don't have the overhead of creating
+        // another trace when we need to materialize `y`.
+        //
+        _ = collectLazyOp(lazyOp)
+        lazyOpsCache.removeAll()
+    }
+
+    var signature:  String {
+        let inputsDesc = inputs.map { input -> String in
+            let dtypeAttr = input.attrs["dtype"]!
+            return "\(input.outputName): \(dtypeAttr)"
+        }
+        let inputDesc = inputsDesc.joined(separator: ", ")
+        let outputsDesc = outputs.map { "\($0.outputName)" }
+        let outputDesc = outputsDesc.joined(separator: ", ")
+        return "lazyTrace_\(operations.count)(\(inputDesc)) -> (\(outputDesc))"
+    }
+
+    private var lazyOpsCache: [ObjectIdentifier: LazyTensorOperation] = [:]
+
+    private func updateCacheAndOperations(
+        _ id: ObjectIdentifier, _ node: LazyTensorOperation) {
+        lazyOpsCache[id] = node
+        node.id = "\(operations.count)"
+        operations.append(node)
+    }
+
+    private func makeConstTensor(with conc: TFETensorHandle) -> LazyTensor {
+        let cTensorHandle = conc._cTensorHandle
+        let result = LazyTensorOperation("Const", 1)
+        let dtype = TensorDataType(TFE_TensorHandleDataType(cTensorHandle))
+        let dtypeAttr = LazyTensorOperation.Attribute.TensorDataTypeValue(dtype)
+        result.attrs = [
+            "dtype": dtypeAttr,
+            "value": LazyTensorOperation.Attribute.ConstTensor(conc)]
+        updateCacheAndOperations(ObjectIdentifier(conc), result)
+        return LazyTensor(_lazy: result, index: 0)
+    }
+
+    private func makePlaceholderTensor(
+        with conc: TFETensorHandle
+    ) -> LazyTensor {
+        let cTensorHandle = conc._cTensorHandle
+        let dtype = TensorDataType(TFE_TensorHandleDataType(cTensorHandle))
+        let dtypeAttr = LazyTensorOperation.Attribute.TensorDataTypeValue(dtype)
+        let placeholder = LazyTensorOperation("Placeholder", 1)
+        placeholder.attrs = ["dtype": dtypeAttr]
+        updateCacheAndOperations(ObjectIdentifier(conc), placeholder)
+        inputs.append(placeholder)
+        inputValues.append(conc)
+        return LazyTensor(_lazy: placeholder, index: 0)
+    }
+
+    private func makeConstTensorOrPlaceholder(
+        with conc: TFETensorHandle, asConst: Bool
```

Comment:
`conc` is neither a widely accepted term of art, nor a word that's explained anywhere in this file. I'd suggest using the full name instead of abbreviations, which is generally what Swift recommends. Same for all occurrences of `conc`.

Reply:
used `handle` everywhere.